app.py

```diff
@@ -0,0 +1,161 @@
+
+# Retry configuration for connecting to the RPC endpoint
+@retry(
+    stop=stop_after_attempt(5),
+    wait=wait_exponential(multiplier=1, min=4, max=10),
+    before_sleep=before_sleep_log(logger, logging.INFO)
+)
+def connect_rpc(url):
+    web3 = Web3(Web3.HTTPProvider(url))
+    if web3.is_connected():
+        logger.info(f"Successfully connected to the RPC endpoint: {url}")
+        return web3
+    else:
+        raise ConnectionError(f"Failed to connect to the RPC endpoint: {url}")
+
+def send_transaction(web3, private_key):
+    account = web3.eth.account.from_key(private_key)
+    try:
+        chain_id = web3.eth.chain_id
+        nonce = web3.eth.get_transaction_count(account.address)
+
+        transaction = {
+            'from': account.address,
+            'to': account.address,  # Sending to self
+            'value': web3.to_wei(0.01, 'ether'),
+            'nonce': nonce,
+            'chainId': chain_id
+        }
+
+        # Estimate gas cost
+        gas = web3.eth.estimate_gas(transaction)
+        transaction['gas'] = gas
+        transaction['gasPrice'] = web3.eth.gas_price
+
+        signed_txn = web3.eth.account.sign_transaction(transaction, private_key)
+        tx_hash = web3.eth.send_raw_transaction(signed_txn.rawTransaction)
+        receipt = web3.eth.wait_for_transaction_receipt(tx_hash)
+
+        if receipt.status == 1:
+            logger.info(f"Transaction successful: {tx_hash.hex()} with estimated gas: {gas}")
+            return True
+        else:
+            logger.error(f"Transaction failed: {tx_hash.hex()}")
+            return False
+    except Exception as e:
+        logger.error(f"Transaction error: {str(e)}")
+        return False
+
+# Background task to periodically send transactions
+def start_transaction_task(web3, private_key, interval):
+    def task():
+        while True:
+            success = send_transaction(web3, private_key)
+            transaction_success_metric.set(1 if success else 0)
+            time.sleep(interval)
+
+    if private_key:
+        thread = threading.Thread(target=task)
+        thread.daemon = True
+        transaction_success_metric = Gauge('gelato_transaction_success', 'Indicates if the last transaction was successful', registry=registry)
+        thread.start()
+    else:
+        logger.info("No PRIVATE_KEY provided. Transactions will not be sent.")
+
+rpc_url, addresses_to_monitor, private_key, transaction_interval, port = initialize()
+web3 = connect_rpc(rpc_url)
+
+app = Flask(__name__)
+
+# Initialize Prometheus metrics
+block_height_metric = Gauge('gelato_block_height', 'The latest block height', registry=registry)
+
+def collect_metrics():
+    latest_block_height = web3.eth.block_number
+    block_height_metric.set(latest_block_height)
+
+    if addresses_to_monitor:
+        balance_metrics = Gauge('gelato_balance', 'Balance of Ethereum addresses', ['address', 'name'], registry=registry)
```

Comment:
Is it OK to recreate Gauge every time? 🤔 Usually they are reused as they collect data inside. Not sure how it's implemented here, but I'd better move all `*_metrics` to global variables.